code/datums/json_savefile.dm

```diff
@@ -79,3 +81,49 @@ GENERAL_PROTECT_DATUM(/datum/json_savefile)
 				continue
 			READ_FILE(savefile, entry_value) //we are cd'ed to the entry, so we don't need to specify a path to read from
 			region[entry] = entry_value
+
+/// Proc that handles generating a JSON file (prettified if 515 and over!) of a user's preferences and showing it to them.
+/// Requester is passed in to the ftp() and tgui_alert() procs, and account_name is just used to generate the filename.
+/// We don't _need_ to pass in account_name since this is reliant on the json_savefile datum already knowing what we correspond to, but it's here to help people keep track of their stuff.
+/datum/json_savefile/proc/export_json_to_client(mob/requester, account_name)
+	if(!istype(requester) || !path)
+		return
+
+	if(!json_export_checks(requester))
+		return
+
+	COOLDOWN_START(src, download_cooldown, (CONFIG_GET(number/seconds_cooldown_for_preferences_export) SECONDS))
+	var/file_name = "[account_name ? "[account_name]_" : ""]preferences_[time2text(world.timeofday, "MMM_DD_YYYY_hh-mm-ss")].json"
+	var/temporary_file_storage = "data/preferences_export_working_directory/[file_name]"
+	var/exportable_json
+
+#if DM_VERSION >= 515
+	if(text2file(json_encode(tree, JSON_PRETTY_PRINT), temporary_file_storage))
+		exportable_json = file(temporary_file_storage)
+#else
+	if(text2file(json_encode(tree), temporary_file_storage))
+		exportable_json = file(temporary_file_storage)
+#endif
+
+	if(!isfile(exportable_json))
+		tgui_alert(requester, "Failed to export preferences to JSON! You might need to try again later.", "Export Preferences JSON")
+		return
```

Comment:
Style nit, especially after we stop supporting 514 for building, this should look something like:

```
if (!text2file(whatever))
    tgui_alert(requester, "oh no")
    return

var/exportable_json = file(temporary_file_storage)
```

It's okay to copy and paste the tgui_alert since the duplicate will not exist for long.